extensions/aws/processors/ListS3.cpp

```diff
@@ -0,0 +1,294 @@
+}
+
+ListS3::ListingState ListS3::getCurrentState(const std::shared_ptr<core::ProcessContext> &context) {
+  ListS3::ListingState current_listing_state;
+  std::unordered_map<std::string, std::string> state;
+  if (!state_manager_->get(state)) {
+    logger_->log_info("No stored state for listed objects was found");
+    return current_listing_state;
+  }
+
+  current_listing_state.listed_key_timestamp = getLatestListedKeyTimestamp(state);
+  logger_->log_debug("Restored previous listed timestamp %lld", current_listing_state.listed_key_timestamp);
+
+  current_listing_state.listed_keys = getLatestListedKeys(state);
+  return current_listing_state;
+}
+
+void ListS3::storeState(const std::shared_ptr<core::ProcessContext> &context, const ListS3::ListingState &latest_listing_state) {
+  std::unordered_map<std::string, std::string> state;
+  state[LATEST_LISTED_KEY_TIMESTAMP] = std::to_string(latest_listing_state.listed_key_timestamp);
+  for (std::size_t i = 0; i < latest_listing_state.listed_keys.size(); ++i) {
+    state[LATEST_LISTED_KEY_PREFIX + std::to_string(i)] = latest_listing_state.listed_keys.at(i);
+  }
+  logger_->log_debug("Stored new listed timestamp %lld", latest_listing_state.listed_key_timestamp);
+  state_manager_->set(state);
+}
+
+void ListS3::createNewFlowFile(
+    const std::shared_ptr<core::ProcessSession> &session,
+    const aws::s3::ListedObjectAttributes &object_attributes) {
+  auto flow_file = session->create();
+  session->putAttribute(flow_file, "s3.bucket", list_request_params_.bucket);
+  session->putAttribute(flow_file, core::SpecialFlowAttribute::FILENAME, object_attributes.filename);
+  session->putAttribute(flow_file, "s3.etag", object_attributes.etag);
+  session->putAttribute(flow_file, "s3.isLatest", object_attributes.is_latest ? "true" : "false");
+  session->putAttribute(flow_file, "s3.lastModified", std::to_string(object_attributes.last_modified));
+  session->putAttribute(flow_file, "s3.length", std::to_string(object_attributes.length));
+  session->putAttribute(flow_file, "s3.storeClass", object_attributes.store_class);
+  if (!object_attributes.version.empty()) {
+    session->putAttribute(flow_file, "s3.version", object_attributes.version);
+  }
+  writeObjectTags(list_request_params_.bucket, object_attributes, session, flow_file);
+  writeUserMetadata(object_attributes, session, flow_file);
+
+  session->transfer(flow_file, Success);
+}
+
+void ListS3::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  logger_->log_debug("ListS3 onTrigger");
+
+  auto aws_results = s3_wrapper_.listBucket(list_request_params_);
+  if (!aws_results) {
+    logger_->log_error("Failed to list S3 bucket %s", list_request_params_.bucket);
+    context->yield();
+    return;
+  }
+
+  if (aws_results->size() == 0) {
+    logger_->log_debug("No S3 object were found in bucket %s", list_request_params_.bucket);
+    return;
```

Comment:
CWEL yields when there are no events, GetFile does not, should we yield here?

Reply:
I'm not sure, even though we could yield if we did not get any data just to wait some more time for additional objects, we cannot be sure, that the next schedule time will not provide new data.

Reply:
I checked the Nifi implementation for reference and it seems it yields in case there is no new S3 object to list. I'll update the implementation accordingly.

Reply:
Updated in [8216738](https://github.com/apache/nifi-minifi-cpp/pull/975/commits/8216738921e0d5368e532b09aaf618fc61bcbaaf)